File: kanban-frontend/src/app/services/kanban.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Board } from '../models/board.model';
import { Column } from '../models/column.model';
import { Card } from '../models/card.model';
import {MoveCardRequest} from '../models/mode-card-request.model';

@Injectable({
  providedIn: 'root'
})
export class KanbanService {
  private http = inject(HttpClient);
  private readonly apiUrl = 'http://localhost:7071/api';

  // ========================
  // BOARD
  // ========================

  getBoards(): Observable<Board[]> {
    return this.http.get<Board[]>(`${this.apiUrl}/boards`);
  }

  getBoardById(id: string): Observable<Board> {
    return this.http.get<Board>(`${this.apiUrl}/boards/${id}`);
  }

  createBoard(board: Partial<Board>): Observable<Board> {
    return this.http.post<Board>(`${this.apiUrl}/boards`, board);
  }

  updateBoard(id: string, board: Board): Observable<Board> {
    return this.http.put<Board>(`${this.apiUrl}/boards/${id}`, board);
  }

  deleteBoard(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/boards/${id}`);
  }

  // ========================
  // COLUMNS
  // ========================

  addColumn(boardId: string, column: Partial<Column>): Observable<Column> {
    return this.http.post<Column>(`${this.apiUrl}/boards/${boardId}/columns`, column);
  }

  updateColumn(boardId: string, columnId: string, column: Column): Observable<Column> {
    return this.http.put<Column>(`${this.apiUrl}/boards/${boardId}/columns/${columnId}`, column);
  }

  deleteColumn(boardId: string, columnId: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/boards/${boardId}/columns/${columnId}`);
  }

  // ========================
  // CARDS
  // ========================

  addCard(boardId: string, columnId: string, card: Partial<Card>): Observable<Card> {
    return this.http.post<Card>(`${this.apiUrl}/boards/${boardId}/columns/${columnId}/cards`, card);
  }

  updateCard(boardId: string, columnId: string, cardId: string, card: Card): Observable<Card> {
    return this.http.put<Card>(`${this.apiUrl}/boards/${boardId}/columns/${columnId}/cards/${cardId}`, card);
  }

  deleteCard(boardId: string, columnId: string, cardId: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/boards/${boardId}/columns/${columnId}/cards/${cardId}`);
  }

  moveCard(boardId: string, request: MoveCardRequest): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/boards/${boardId}/move-card`, request);
  }
}

File: Kanban.Functions/Services/KanbanService.cs
using Kanban_Functions.Models;
using Kanban_Functions.Services.Interfaces;
using MongoDB.Driver;

namespace Kanban_Functions.Services;

public class KanbanService: IKanbanService
{
    private readonly IMongoCollection<BoardModel> _boards;

    public KanbanService(MongoDbContext context)
    {
        _boards = context.Boards;
    }

    // Boards
    
    public async Task CreateBoard(BoardModel board)
    {
        await _boards.InsertOneAsync(board);
    }

    public async  Task UpdateBoard(string id, BoardModel board)
    {
        await  _boards.ReplaceOneAsync(board => board.Id == id, board);
    }

    public async  Task DeleteBoard(string id)
    {
        await  _boards.DeleteOneAsync(board => board.Id == id);
    }

    public async Task<List<BoardModel>> GetAllBoards()
    {
        return await _boards.Find(board => true).ToListAsync();
    }

    public async Task<BoardModel> GetBoardById(string id)
    {
        return await _boards.Find(board => board.Id == id).FirstOrDefaultAsync();
    }
 
    // Columns
    
    public async  Task CreateColumn(string boardId, ColumnModel column)
    {
        var board = await GetBoardById(boardId);
        if (board == null) return;
        
        board.Columns.Add(column);
        await UpdateBoard(boardId, board);
    }

    public async Task UpdateColumn(string boardId, string columnId, ColumnModel column)
    {
        var board = await GetBoardById(boardId);
        if (board == null) return;
        
        var index = board.Columns.FindIndex(column => column.Id == columnId);
        if (index != -1)
        {
            column.Cards = board.Columns[index].Cards;
            board.Columns[index] = column;
            
            await UpdateBoard(boardId, board);
        }
    }

    public async Task DeleteColumn(string boardId, string columnId)
    {
        var board = await GetBoardById(boardId);
        if (board == null) return;
        
        board.Columns.RemoveAll(column => column.Id == columnId);
        await UpdateBoard(boardId, board);
    }

    // Cards
    
    public async Task CreateCard(string boardId, string columnId, CardModel card)
    {
        var board = await GetBoardById(boardId);
        if (board == null) return;
        
        var column = board.Columns.FirstOrDefault(column => column.Id == columnId);
        if (column != null)
        {
            column.Cards.Add(card);
            await UpdateBoard(boardId, board);
        }
    }

    public async Task UpdateCard(string boardId, string columnId, string cardId, CardModel card)
    {
        var board = await GetBoardById(boardId);
        if (board == null) return;
        
        var column = board.Columns.FirstOrDefault(column => column.Id == columnId);
        if (column != null)
        {
            var cardIndex = column.Cards.FindIndex(card => card.Id == cardId);
            if (cardIndex != -1)
            {
                column.Cards[cardIndex] = card;
                await UpdateBoard(boardId, board);
            }
        }
    }

    public async Task DeleteCard(string boardId, string columnId, string cardId)
    {
        var board = await GetBoardById(boardId);
        if (board == null) return;
        
        var column = board.Columns.FirstOrDefault(column => column.Id == columnId);
        if (column != null)
        {
            column.Cards.RemoveAll(card => card.Id == cardId);
            await UpdateBoard(boardId, board);
        }
        
    }

    public async Task MoveCard(string boardId, string sourceColumnId, string targetColumnId, string cardId)
    {
        var board = await GetBoardById(boardId);
        if (board == null) return;
        
        var sourceCol = board.Columns.FirstOrDefault(c => c.Id == sourceColumnId);
        var targetCol = board.Columns.FirstOrDefault(c => c.Id == targetColumnId);

        if (sourceCol != null && targetCol != null)
        {
            var card = sourceCol.Cards.FirstOrDefault(card => card.Id == cardId);

            if (card != null)
            {
                sourceCol.Cards.Remove(card);
                targetCol.Cards.Add(card);
                await UpdateBoard(boardId, board);
            }
        }
    }
}